Fixes recursive dict encoding

matt035343 opened the pull request:
Fixes #391, please refer to this issue for a detailed description of what is fixed in this pull request.

I implemented both a fix to the encoding and added tests that will capture the behaviour.

AbsurdlySuspicious commented:
Wouldn't this still fail on dicts or lists with more than one level of nesting? I believe this type check should be done in `_encode_json_type` to be properly recursive

matt035343 commented:
@AbsurdlySuspicious

> Wouldn't this still fail on dicts and lists with more than one level of nesting? I believe this type check should be done in `_encode_json_type` to be properly recursive

You are absolutely right. I have now fixed it and changed the test to catch this case as well.

george-zubrienko commented:
@matt035343 could you sync your branch with master, so CI runs again w/o py3.6?
